### ISSUE TITLE
std::in_place_stop_source::get_token and std::get_stop_token const correctness

### ISSUE THREAD

**Issue by ericniebler** [cplusplus/sender-receiver#31]
<a href="https://github.com/jtsylve"><img src="https://avatars.githubusercontent.com/u/594165?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [jtsylve](https://github.com/jtsylve)**
_Thursday Aug 17, 2023 at 20:39 GMT_
_Originally opened as https://github.com/brycelelbach/wg21_p2300_execution/issues/46_

----

In the current proposal, `in_place_stop_source::get_token` is `const` qualified, and `std::get_stop_token`  requires a const parameter.

The problem in practice is that token callbacks almost certainly need to register themselves with the source, which tends to require that the token maintain a pointer or reference to the source.  With `get_token` being const qualified, the source can only pass a constant reference or pointer to itself to be stored by the token, but the callback almost certainly needs to mutate the source to register itself.

`libunifex` doesn't have this issue because their implementation of `get_token` [is not const qualified](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/inplace_stop_token.hpp#L90).
`stdexec` follows the specification and had to resort to [marking its list of callbacks as `mutable`](https://github.com/NVIDIA/stdexec/blob/48c52df0f81c6151eecf4f39fa5eed2dc0216204/include/stdexec/stop_token.hpp#L153), which is usually considered a code smell by many.

I am curious about the reasoning behind this decision.  I know that `std::stop_source` made the same choice, but those implementations do not require callback registration.

**Comment by lewissbaker** [comment]
The `get_token()` method on the `in_place_stop_source` is const because the operation does not logically modify the visible state of the stop-source, either immediately or allow subsequent modification through the returned stop-token. The only operation that can affect the visible state is a call to `request_stop()` which is non-const.

Yes, this means you will probably need to use one or more mutable members in the stop-source class to manage the registered stop-callbacks. I don't see this as a problem - that's what the `mutable` keyword is for - implementing logically const operations that do actually need to mutate the state of the object.

**Comment by jensmaurer** [comment]
I'd like to point out that "const" member functions are thread-safe per the standard library rules. Can multiple threads call `get_token()` concurrently?

**Comment by lewissbaker** [comment]
Yes, it is safe to call all member functions of `in_place_stop_source` concurrently (except for the destructor).

In practice, `get_token()` just captures a pointer to some internal state of the stop source and does not perform any mutations. The mutations that are performed through the token occur when constructing a stop callback, which is safe to do concurrently, provided the return from the stop callback destructors happen before the associated stop source destructor is called.
